### PR TITLE
Refactor num_samples / temperature, rename elasticity to localization, add warnings, init_loss before sampling, OnlineLLC fix

### PR DESCRIPTION
**Pull request by svwingerden**
This PR includes a few changes:
1. I refactored the usage of `num_samples` and `temperature` in SGLD (and SGNHT). We now only pass along a temperature, which if not set defaults to the optimal temperature.
2. I renamed the `elasticity` parameter in SGLD to `localization`.
3. I added a few warnings (burnin, weird temperature)
4. I fixed `OnlineLLCEstimator` not to return an average of a moving average
5. If init_loss is not supplied to `sample()` or any of the `estimate()` functions, we calculate it explicitly on one batch. This allows burn-in to work, btw.
6. Refactored some tests, added in a new seeding test

### CODE REVIEW

File: src/devinterp/slt/llc.py [hunk 1]
```diff
@@ -6,106 +6,120 @@
...
 
-    def update(self, chain: int, draw: int, loss: float):
-        self.losses[chain, draw] = loss 
-
-    @property
-    def init_loss(self):
-        return self.losses[0, 0]
+    def update(self, chain: int, draw: int, loss: float, init_loss):
+        self.losses[chain, draw] = loss
+        self.init_loss = init_loss  # This is clunky, sorry
```
Comment (georgeyw) [inline src/devinterp/slt/llc.py:53]: what's the reason for changing this? (not saying the other way was graceful, just curious if e.g. this has any behavioral change or is useful for some other reason)

File: src/devinterp/slt/llc.py [hunk 2]
```diff
@@ -6,106 +6,120 @@
...
         num_draws (int): Number of samples to draw. (should be identical to param passed to sample())
-        n (int): Number of samples used to calculate the LLC.
+        temperature (float): Temperature used to calculate the LLC.
         device (Union[torch.device, str]): Device to perform computations on, e.g., 'cpu' or 'cuda'.
     """
-    def __init__(self, num_chains: int, num_draws: int, n: int, device="cpu"):
+
+    def __init__(
+        self, num_chains: int, num_draws: int, temperature: float, device="cpu"
+    ):
```
Comment (georgeyw) [inline src/devinterp/slt/llc.py:90]: just leaving this comment here, but it's a broader request: if there's going to be lots of files touched by both a refactor/feature update + linting, and if it's not too much trouble, can the linting be split out into a different PR? just so that it's easier to know which lines of change need what level of attention

File: src/devinterp/slt/llc.py [hunk 3]
```diff
@@ -6,106 +6,120 @@
...
         self.device = device
 
-    def update(self, chain: int, draw: int, loss: float):
-        self.losses[chain, draw] = loss 
-        init_loss = self.losses[chain, 0]
-
-        if draw == 0:  # TODO: We can probably drop this and it still works (but harder to read)
-            self.llcs[chain, draw] = 0.
+    def update(self, chain: int, draw: int, loss: float, init_loss: float):
+        self.init_loss = init_loss
```
Comment (georgeyw) [inline src/devinterp/slt/llc.py:110]: not sure if this is accurate since I haven't read the rest of the code yet, but I'm getting the sense that maybe the `init_loss` passed here is computed outside of the estimator (e.g. using the recorded training loss for a checkpoint). if that's the case, maybe it makes sense to pass it during the llc estimator initialization and set it as a class variable there instead of initializing it at the update step?

File: src/devinterp/slt/llc.py [hunk 4]
```diff
@@ -115,19 +129,19 @@ def sample(self):
             "llc/means": self.llc_means.cpu().numpy(),
             "llc/stds": self.llc_stds.cpu().numpy(),
             "llc/trace": self.llcs.cpu().numpy(),
```
Comment (georgeyw) [inline src/devinterp/slt/llc.py:131]: since `self.moving_avg_llcs` is taking the place of `self.llcs`, do we want to return the former here or is the behavioral change w/ returning the latter intended?

File: src/devinterp/slt/llc.py [hunk 5]
```diff
@@ -115,19 +129,19 @@ def sample(self):
             "llc/means": self.llc_means.cpu().numpy(),
```
Comment (georgeyw) [inline src/devinterp/slt/llc.py:129]: i think i also just noticed that this should probably be `llc/mean` and not `llc/means` to be in line with the not online llc estimator above? also same w/ `llc/stds -> llc/std`

File: src/devinterp/optim/sgnht.py [hunk 6]
```diff
@@ -37,21 +39,24 @@ def __init__(
         :param lr: Learning rate
         :param diffusion_factor: The diffusion factor of the thermostat (default: 0.01)
         :param bounding_box_size: the size of the bounding box enclosing our trajectory The diffusion factor (default: None)
-        :param num_samples: Number of samples to average over (default: 1)
+        :param temperature: Temperature, float (default: 1., overridden to be num_samples * np.log(num_samples))
```
[the remark this answers is not in the document]
Comment (georgeyw) [reply src/devinterp/optim/sgnht.py:42]: ok i see we do this later on with e.g. `estimate_learning_coefficient_with_summary`, but maybe it wouldn't hurt just to have it default to it here as well

File: src/devinterp/utils.py [hunk 7]
```diff
@@ -1,31 +1,91 @@
...
+    plt.legend(loc="upper left", bbox_to_anchor=(1, 1))
     plt.title(title)
     plt.xlabel(x_axis)
     plt.ylabel(y_axis)
     plt.figure(figsize=fig_size)
     plt.tight_layout()
     plt.show()
+
+
+def optimal_temperature(data):
```
Comment (georgeyw) [inline src/devinterp/utils.py:55]: i think in practice, jesse and i both used the batch size for `num_samples`, `len(data.dataset)` was too large (and also didn't make sense to me). `batch_size` makes sense to me since varying the batch size does change the effective temperature of SGLD steps, but you should be able to e.g. run sampling using a dataset of 1m rows where you don't use all the rows and get the same llc estimate as a counterfactual sampling run using a dataset of 100m rows where you also don't use all the rows (but do the same batch size, num draws, num chains, etc)

File: src/devinterp/utils.py [hunk 8]
```diff
@@ -1,31 +1,91 @@
...
+        return len(data.dataset) / np.log(len(data.dataset))
+    elif isinstance(data, Dataset) or isinstance(data, list):
+        return len(data) / np.log(len(data))
+    elif isinstance(data, int):
+        return data / np.log(data)
+    else:
+        raise NotImplementedError(f'Temperature for data type {type(data)} not implemented, use Dataloader or Dataset instead.')
+
+
+def get_init_loss_one_batch(dataloader, model, criterion, device):
```
Comment (georgeyw) [inline src/devinterp/utils.py:66]: in the prev iteration of llc estimation, we used the average minibatch loss over the num chains, which was neither 1 batch nor the full dataset loss. this seemed to work pretty well, so maybe it's worth changing this (or creating a third util) to average over `k` batches?

File: src/devinterp/utils.py [hunk 9]
```diff
@@ -1,31 +1,91 @@
...
+    model.train()
+    loss = 0.
+    with torch.no_grad():
+        for xs, ys in dataloader:
+            xs, ys = xs.to(device), ys.to(device)
+            y_preds = model(xs)
+            loss += criterion(y_preds, ys).detach().item()
+    return loss
+
+def transformers_cross_entropy(inputs, outputs):
```
Comment (georgeyw) [inline src/devinterp/utils.py:88]: haven't looked at the notebooks yet, but if this is meant to be a utility loss function to be passed as the criterion for things like llc estimation or `get_init_loss` functions, then i've been using this implementation (for training as well): https://github.com/neelnanda-io/TransformerLens/blob/main/transformer_lens/utils.py#L121

in particular i think at minimum you need the token offset between the logits and the outputs

**Review by georgeyw — CHANGES_REQUESTED** [review]
left some comments / questions in the code. did not look at notebooks or tests yet, will look at those on a second round of review once we have a back and forth on the current comments

File: src/devinterp/utils.py (hunk 9 above)
Comment (georgeyw) [reply src/devinterp/utils.py:88]: nah i think it's fine to delete it, we can defer the decision about including architecture-specific utils for another day

**Review by georgeyw — APPROVED** [review]
Added a few more comments, mostly nitpicks

File: src/devinterp/slt/sampler.py [hunk 10]
```diff
@@ -162,3 +195,101 @@ def get_args(i):
             callback.finalize()
 
 
+def estimate_learning_coeff_with_summary(
```
[the remark this answers is not in the document]
Comment (georgeyw) [reply src/devinterp/slt/sampler.py:198]: (e.g. this breaks an import in a notebook (slgd calibration) in the mala PR branch)
